File: lc_soln/0016_3sum_closest/solution.cpp
#include <algorithm>
#include <climits>
#include <vector>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        ranges::sort(nums);
        int result = 0;
        int min_diff = INT_MAX;
        for (int i = 0; i < nums.size() - 2; ++i) {
            int l = i + 1;
            int r = nums.size() - 1;
            while (l < r) {
                int total = nums[i] + nums[l] + nums[r];
                if (total == target) {
                    return total;
                }
                if (int diff = abs(total - target); diff < min_diff) {
                    min_diff = diff;
                    result = total;
                }
                if (total > target) {
                    r--;
                } else {
                    l++;
                }
            }
        }
        return result;
    }
};

File: lc_soln/0016_3sum_closest/solution.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums = sorted(nums)
        result = 0
        min_diff = float("inf")
        n = len(nums)
        for i in range(n):
            l = i + 1
            r = n - 1
            while l < r:
                total = nums[i] + nums[l] + nums[r]
                if total == target:
                    return total
                diff = abs(total - target)
                if diff < min_diff:
                    min_diff = diff
                    result = total
                if total > target:
                    r -= 1
                else:
                    l += 1
        return result
